Ensure pods both running and ready before starting e2e tests

mbforbes opened the pull request:
This pushes the problem we're seeing in #8378 with the reboot test to a much higher level.

Before the e2e tests run, we were previously checking that all pods are "running" by ensuring:
`pod.Status.Phase == api.PodRunning`

This PR makes it so that we also check that pods are "ready" by ensuring:
`pod.Status.Conditions ⊇ {Type: api.PodReady, Status: api.ConditionTrue}`

The reboot test was doing this more thorough checking as a precondition check, which was failing.

Oh, and this also rips out some shell in favor of Go :-) (chatting with @ixdy there's no other place to put this code than in `driver.go` right now).

mbforbes committed:
Ensure pods both running and ready before starting e2e tests

mbforbes commented:
+cc @quinton-hoole 
+cc @roberthbailey 
+cc @ixdy

Any of you three care to review?

Running e2es now.

roberthbailey commented on hack/e2e-internal/e2e-status.sh:
```diff
@@ -23,7 +23,7 @@ set -o pipefail
 
 : ${KUBE_VERSION_ROOT:=${KUBE_ROOT}}
-: ${KUBECTL:="${KUBE_VERSION_ROOT}/cluster/kubectl.sh"}
+: ${KUBECTL:=${KUBE_VERSION_ROOT}/cluster/kubectl.sh}
```
Why did you get rid of the quotes?

mbforbes replied:
They aren't necessary and break github's code highlighting (see current version of file: https://github.com/mbforbes/kubernetes/blob/master/hack/e2e-internal/e2e-status.sh)

quinton-hoole replied:
Unless of course KUBE_VERSION_ROOT happens to have a space in it :-)

mbforbes replied:
But KUBE_VERSION_ROOT, which is declared on the line above, doesn't use quotes in its definition.

zmerlynn replied:
Please either make this consistent across all of e2e-internal or revert.

mbforbes replied:
Made consistent.

roberthbailey commented on test/e2e/driver.go:
```diff
@@ -31,6 +33,18 @@ import (
 
+const (
+	// podStartupTimeout is the time to allow all pods in the cluster to become
+	// running and ready before any e2e tests run. It includes pulling all of
```
pulling or polling?

mbforbes replied:
pulling

roberthbailey commented:
Seems reasonable. Assigning to @ixdy for final review.

ixdy commented:
Not sure if we want to try to merge this before or after #8428 (which is now ready for merge). Whoever loses the race will have to resolve some conflicts.

mbforbes commented:
Merge yours first; I don't mind resolving conflicts, and you've been trying on that one for longer :-)

ixdy commented:
My change merged, so you'll need to resolve conflicts. Most notably, test/e2e/driver.go (mostly) turned into test/e2e/e2e_test.go.

mbforbes commented:
e2es passed with GCE, but going to running again (now) after merge.

In case you'd like to take a look again before this is finally in, my changes since last review:

- merged (duh)
- added more quotes around pods and namespaces in e2e's util.go for clarity
- decreased startup pod "running and ready" timeout from 20m to 10m. It was taking 40s on my laptop, so this seems reasonable—it will cut the time for us to see this failed in half.

quinton-hoole commented on test/e2e/e2e_test.go:
```diff
@@ -33,6 +35,18 @@ import (
+	// podStartupTimeout is the time to allow all pods in the cluster to become
+	// running and ready before any e2e tests run. It includes pulling all of
+	// the pods (as of 5/18/15 this is 8 pods).
+	podStartupTimeout = 10 * time.Minute
```
I think that we should reduce this further.  Most e2e tests succeed or fail in about a minute.

zmerlynn replied:
Pulling all of the initial pods can easily take 10 minutes.

mbforbes replied:
Yeah, this check is only run once, before the e2e suite starts; also ditto what Zach said.

quinton-hoole commented on test/e2e/util.go:
```diff
@@ -105,6 +105,83 @@ func providerIs(providers ...string) bool {
 
 type podCondition func(pod *api.Pod) (bool, error)
 
+// podReady returns whether pod has a condition of Ready in the status True.
```
Remove last word?

mbforbes replied:
I think it's actually correct. To explain: pod has an array of conditions. One of the conditions it can have is the readiness condition (called "Ready"). This condition can have a status of True or False. We want the ready condition to have status True.

ixdy replied:
maybe replace "in" with "with"?

mbforbes replied:
Done (not yet pushed).

quinton-hoole commented on test/e2e/util.go:
```diff
@@ -105,6 +105,83 @@ func providerIs(providers ...string) bool {
+}
+
+// podRunningReady checks whether pod p's phase is running and it has a ready
+// condition of status true.
```
Remove "condition of"

mbforbes replied:
See comment above, please.

quinton-hoole commented on test/e2e/e2e_test.go:
```diff
@@ -92,6 +106,15 @@ func TestE2E(t *testing.T) {
+	// cluster infrastructure pods that are being pulled or started can block
+	// test pods from running, and tests that ensure all pods are running and
+	// ready will fail).
+	if err := waitForPodsRunningReady(api.NamespaceDefault, minStartupPods, podStartupTimeout); err != nil {
```
I find this rather confusing.  We say we're waiting for all pods to be ready, but in the current configuration we only wait for any one pod to be ready.  Also, should individual tests that require specific pods to be running in order to succeed not perform their own specific checks against the states of those pods.

Otherwise stated, I don't understand how ensuring that one pod is running in the default namespace is good enough to be useful to any tests on a reliable basis?

I'm coming in rather late here, so it's quite possible that I'm missing some context here, in which case please accept my apologies in advance.

zmerlynn replied:
We're waiting for the system pods to be created, pulled and running, but for paranoia you don't want to exit the check way too early and exit when there have been 0 pods created. In practice, `minStartupPods` closes a highly theoretical race, at the expense of increasing confusion.

quinton-hoole replied:
I still don't understand.  What race does it solve?  And how can checking that 1 system pod is ready be a sufficiently reliable proxy for all system pods being ready?

ixdy replied:
The comment is a little confusing, but what this function checks is that
- there is at least one pod
- that all pods are running and ready

The first condition catches the race in which this function runs before any pods are created; in that case, the second condition would be vacuously true.

zmerlynn replied:
It's a proxy to ensure that we've started starting any addons. Once they're actually being created, they're actually submitted to the system super fast, but ***in theory*** `kube-addons` may not have run yet. In practice, it's basically run immediately after the apiserver comes up.

mbforbes replied:
Thanks for chiming in everyone. Would you prefer I update this comment?

ixdy replied:
I think this comment may be fine, but the comment before minStartupPods (L44-46) is very misleading.

mbforbes replied:
Considerably lengthened that comment (not yet pushed).

mbforbes commented:
- e2e passes on gce
- shipping kickable
- running e2es on gke currently (shouldn't solve problems, but should at least run)

mbforbes commented:
GKE e2es pass this new check. Seeing flakes / failures on GKE e2es, but given this runs at startup, I believe these are unrelated. Of course, up to the mergers to decide. Fine from my end post-Shippable.

ixdy commented:
LGTM besides minor comment nits - I'll let the others chime in too, though.

mbforbes commented:
I think this is ready unless... wait, is driver.go back? Agh!

ixdy commented:
... tell me about it. :)

Wait for #8473 to merge (just as soon as Shippable is happy) and things will be back to the way they were.

mbforbes commented:
Just re-merged. Ugh. OK, will wait and re-re-re-re-merge.

mbforbes commented:
Shippable is green—@ixdy, would you like to merge this?
